src/document-click.js

```diff
@@ -358,3 +347,52 @@ export function expandTargetHref_(e, target, urlReplacements) {
   }
   return newHref;
 }
+
+
+/** @typedef {{element: !Element, attributes: Object<string, string>}} */
+let RestoreElementAttributesDef;
+
+
+/**
+ * Removes attributes from elements with matching hash target. This includes:
+ *   *[id=hash] and a[name=hash].
+ * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
+ * @param {string} hash
+ * @return {!Array<!RestoreElementAttributesDef>}}
+ * @private
+ */
+function removeAttrsWithMatchingHash_(ampdoc, hash) {
+  const restoreElementsAttrs = [];
+  const targetElements = ampdoc.getRootNode().querySelectorAll(
+      `#${hash},a[name=${hash}]`) || [];
+  for (let i = 0; i < targetElements.length; i++) {
+    const element = targetElements[i];
+    const attributes = {};
```

Comment:
Both:

1. It's enough to just have `RestoreElementAttributesDef` be `{element, attrName, attrValue}` and thus you can just take care of everything with a single loop.
2. We could even simplify this and only look for a single first element that matches either id search or name search. It seems to me that if there are several competing elements - we could live with that. But, obviously, a more foolproof solution is good as well :)